### PR TITLE
[k8s] Share SSH jump pod

### LINKED ISSUE #2598: [k8s] Share SSH jump pod across users in the same namespace
We currently create a SSH jump pod per user, even if all of them use the same namespace:

https://github.com/skypilot-org/skypilot/blob/b1d8c9d8bfca9d9d696ba3ef07a2f2cdf91b061a/sky/clouds/kubernetes.py#L31

This should instead by based on the kubecontext hash (like we have in `get_current_user_identity`) so that different users sharing the same namespace should use the same SSH jump pod. 

We _may_ also need additionally handle keys and auth from different users on the same ssh jump pod.

### PR DESCRIPTION
**Pull request by kbrgl**
This PR closes #2598 and adds plumbing to share the SSH jump pod across a Kubernetes namespace.

1. Per-user secrets for the user's public SSH key are replaced with a single secret object, `sky-ssh-key`, that is created once and patched when future attempts to setup Kubernetes authentication are made.
2. For initial jump pod creation, the manifest runs `cat /etc/secret-volume/ssh-key-* > ~/.ssh/authorized_keys` to set up the initial user's SSH key.
3. When a new user tries to authenticate, the patched secret is re-mounted at /etc/secret-volume and this change is picked up by the LCM script which reruns `cat /etc/secret-volume/ssh-key-* > ~/.ssh/authorized_keys`.

The LCM script now uses 2 threads to poll: one to manage the lifecycle and terminate the jump pod, and one to reload keys on an interval.

## Testing

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_compatibility_tests.sh`

Tested locally.

cc @romilbhardwaj

### CODE REVIEW

File: sky/utils/kubernetes/ssh_jump_lifecycle_manager.py
Comment (romilbhardwaj) [inline sky/utils/kubernetes/ssh_jump_lifecycle_manager.py:12]:
```suggestion
To allow multiple users to the share the same SSH jump pod, 
this script also reloads SSH keys from the mounted secret volume on an
interval and updates `~/.ssh/authorized_keys`.
```

File: sky/utils/kubernetes/ssh_jump_lifecycle_manager.py
Comment (romilbhardwaj) [inline sky/utils/kubernetes/ssh_jump_lifecycle_manager.py:51]: Nice!

File: sky/utils/kubernetes/ssh_jump_lifecycle_manager.py
Comment (romilbhardwaj) [inline sky/utils/kubernetes/ssh_jump_lifecycle_manager.py:85]: Nit
```suggestion
        sys.stdout.write(f'[SSH Key Reloader][ERROR] Failed to reload SSH keys: {e}\n')
```

File: sky/utils/kubernetes/ssh_jump_lifecycle_manager.py
Comment (romilbhardwaj) [inline sky/utils/kubernetes/ssh_jump_lifecycle_manager.py:134]: Nit
```suggestion
            sys.stdout.write('[Lifecycle][ERROR] Deletion failed. Exiting '
```

**Review by romilbhardwaj — COMMENTED** [review]
Awesome, thanks @kbrgl! Took a skim through the code and it looks good. Left some nits. Will test it and approve if all is good.

File: sky/clouds/kubernetes.py
Comment (kbrgl) [reply sky/clouds/kubernetes.py:30]: Yeah I think that's a good idea!

File: sky/utils/kubernetes/ssh_jump_lifecycle_manager.py
Comment (kbrgl) [reply sky/utils/kubernetes/ssh_jump_lifecycle_manager.py:52]: 🥳

**Review by romilbhardwaj — COMMENTED** [review]
Thanks @kbrgl! Did some manual testing - seems to be working nicely. Backward compatibility has some issues, see comments.

File: sky/utils/kubernetes/ssh_jump_lifecycle_manager.py
Comment (romilbhardwaj) [reply sky/utils/kubernetes/ssh_jump_lifecycle_manager.py:82]: On a related note, I've opened #2901 to find a better solution to this backward compatibility problem.

**Review by romilbhardwaj — COMMENTED** [review]
Thanks @kbrgl! Pushed some nits and merge conflict fixes. Manually tested backward compat, now running smoke tests.

Once everything is ready, will merge this PR and push new container images to our image registry.

**Review by romilbhardwaj — APPROVED** [review]
Thanks @kbrgl! Smoke tests pass. Manually tested backward compatibility. Building and updating our docker image, will merge after that's done.

**Comment by romilbhardwaj** [comment]
* Updated images and build script to use date versioning - https://github.com/skypilot-org/skypilot-catalog/pull/60
* Fixed #3158 
* Added socat and netcat to base image on user request for faster provisioning

- [ ] Running k8s smoke tests

**Pull request merged**
